File: fighter.py
import random
from events import AttackArgs, AttackRollArgs, HitArgs, MissArgs
from target import Target
from util import magic_weapon, roll_dice
from feats import (
    GreatWeaponMaster,
    AttackAction,
    ASI,
    PolearmMaster,
    Feat,
    EquipWeapon,
)
from character import Character
from weapons import Glaive, Greatsword, GlaiveButt, Maul
from log import log


class StudiedAttacks(Feat):
    def __init__(self) -> None:
        self.name = "StudiedAttacks"
        self.enabled = False

    def roll_attack(self, args):
        if self.enabled:
            args.adv = True

    def hit(self, args):
        self.enabled = False

    def miss(self, args):
        self.enabled = True


class HeroicAdvantage(Feat):
    def __init__(self):
        self.name = "HeroicAdvantage"

    def begin_turn(self, target):
        self.used = False

    def roll_attack(self, args):
        if self.used or args.adv:
            return
        if args.disadv:
            roll = min(args.roll1, args.roll2)
            if roll < 8:
                self.used = True
                self.adv = True
                args.roll1 = random.randint(1, 20)
        else:
            roll = args.roll1
            if roll < 8:
                self.used = True
                args.adv = True


class ActionSurge(Feat):
    def __init__(self, max_surges) -> None:
        self.name = "ActionSurge"
        self.max_surges = max_surges

    def apply(self, character):
        self.character = character

    def begin_turn(self, target):
        if self.surges > 0:
            self.character.actions += 1
            self.surges -= 1

    def short_rest(self):
        self.surges = self.max_surges


class PrecisionAttack(Feat):
    def __init__(self, low=5) -> None:
        self.name = "PrecisionAttack"
        self.low = low

    def roll_attack(self, args: AttackRollArgs):
        maneuvers = self.character.feat("Maneuvers")
        if maneuvers.used_maneuver:
            return
        if not args.hits() and args.roll() >= self.low:
            roll = maneuvers.roll()
            args.situational_bonus += roll


class TrippingAttack(Feat):
    def __init__(self) -> None:
        self.name = "TrippingAttack"

    def hit(self, args: HitArgs):
        maneuvers = self.character.feat("Maneuvers")
        if maneuvers.used_maneuver:
            return
        if args.target.prone:
            return
        roll = maneuvers.roll()
        if roll > 0:
            args.add_damage("TrippingAttack", roll)
            if not args.target.save(self.character.dc("str")):
                args.target.prone = True


class Maneuvers(Feat):
    def __init__(self, level) -> None:
        self.name = "Maneuvers"
        if level >= 15:
            self.max_dice = 6
        elif level >= 7:
            self.max_dice = 5
        else:
            self.max_dice = 4
        if level >= 18:
            self.superiority_size = 12
        elif level >= 10:
            self.superiority_size = 10
        else:
            self.superiority_size = 8
        self.enabled_relentless = level >= 15
        self.used_maneuver = False
        self.superiority_dice = 0

    def apply(self, character):
        self.character = character

    def short_rest(self):
        self.superiority_dice = self.max_dice

    def begin_turn(self, target: Target):
        self.used_relentless = False

    def before_attack(self):
        self.used_maneuver = False

    def roll(self):
        if self.superiority_dice > 0:
            self.used_maneuver = True
            self.superiority_dice -= 1
            return roll_dice(1, self.superiority_size)
        elif self.enabled_relentless and self.used_relentless:
            self.used_maneuver = True
            self.used_relentless = True
            return roll_dice(1, 8)
        return 0


class ToppleIfNecessaryAttackAction(Feat):
    def __init__(self, num_attacks, topple_weapon, default_weapon) -> None:
        self.name = "ToppleIfNecessaryAttackAction"
        self.topple_weapon = topple_weapon
        self.default_weapon = default_weapon
        self.num_attacks = num_attacks

    def action(self, target: Target):
        for i in range(self.num_attacks):
            weapon = self.default_weapon
            if not target.prone and i < self.num_attacks - 1:
                weapon = self.topple_weapon
            self.character.attack(target, weapon, main_action=True)


class Fighter(Character):
    def __init__(
        self, level, use_pam=False, subclass_feats=[], min_crit=20, use_topple=True
    ):
        base_feats = []
        self.use_pam = use_pam
        self.magic_weapon = magic_weapon(level)
        if use_pam:
            weapon = Glaive(bonus=self.magic_weapon, min_crit=min_crit)
        else:
            weapon = Greatsword(bonus=self.magic_weapon, min_crit=min_crit)
        base_feats.append(EquipWeapon(weapon, savage_attacker=True, max_reroll=2))
        if level >= 20:
            num_attacks = 4
        elif level >= 11:
            num_attacks = 3
        elif level >= 5:
            num_attacks = 2
        else:
            num_attacks = 1
        if use_topple and level >= 5:
            maul = Maul(bonus=self.magic_weapon, min_crit=min_crit)
            base_feats.append(EquipWeapon(maul, savage_attacker=True, max_reroll=2))
            base_feats.append(ToppleIfNecessaryAttackAction(num_attacks, maul, weapon))
        else:
            base_feats.append(AttackAction(attacks=(num_attacks * [weapon])))
        if level >= 13:
            base_feats.append(StudiedAttacks())
        if level >= 17:
            base_feats.append(ActionSurge(2))
        elif level >= 2:
            base_feats.append(ActionSurge(1))
        base_feats.extend(subclass_feats)
        if use_pam:
            butt = GlaiveButt(bonus=self.magic_weapon, min_crit=min_crit)
            base_feats.append(EquipWeapon(butt, savage_attacker=True, max_reroll=2))
            feats = [
                GreatWeaponMaster(),
                PolearmMaster(butt),
                ASI([["str", 1]]),
                ASI(),
                ASI(),
                ASI(),
                ASI(),
            ]
        else:
            feats = [
                GreatWeaponMaster(),
                ASI([["str", 2]]),
                ASI(),
                ASI(),
                ASI(),
                ASI(),
                ASI(),
            ]
        super().init(
            level=level,
            stats=[17, 10, 10, 10, 10, 10],
            feat_schedule=[4, 6, 8, 12, 14, 16, 19],
            feats=feats,
            base_feats=base_feats,
        )


class ChampionFighter(Fighter):
    def __init__(self, level, **kwargs):
        feats = []
        if level >= 10:
            feats.append(HeroicAdvantage())
        if level >= 15:
            min_crit = 18
        elif level >= 3:
            min_crit = 19
        else:
            min_crit = 20
        super().__init__(
            level,
            subclass_feats=feats,
            min_crit=min_crit,
            **kwargs,
        )


class TrippingFighter(Fighter):
    def __init__(self, level, **kwargs):
        feats = []
        if level >= 3:
            feats.append(Maneuvers(level))
            feats.append(TrippingAttack())
        super().__init__(level, subclass_feats=feats, **kwargs)


class BattlemasterFighter(Fighter):
    def __init__(self, level, **kwargs):
        feats = []
        if level >= 3:
            feats.append(Maneuvers(level))
        super().__init__(level, subclass_feats=feats, **kwargs)


class PrecisionFighter(Fighter):
    def __init__(self, level, low=8, **kwargs):
        feats = []
        if level >= 3:
            feats.append(Maneuvers(level))
            feats.append(PrecisionAttack(low=low))
        super().__init__(level, subclass_feats=feats, **kwargs)


class PrecisionTrippingFighter(Fighter):
    def __init__(self, level, low=1, **kwargs):
        feats = []
        if level >= 3:
            feats.append(Maneuvers(level))
            feats.append(TrippingAttack())
            feats.append(PrecisionAttack(low=low))
        super().__init__(level, subclass_feats=feats, **kwargs)

File: dndsim/src/classes/Fighter.ts
import { AbilityScoreImprovement } from "../feats/general/AbilityScoreImprovement"
import { ExtraAttack } from "../feats/shared/ExtraAttack"
import { GreatWeaponFighting } from "../feats/fightingStyle/GreatWeaponFighting"
import { GreatWeaponMaster } from "../feats/general/GreatWeaponMaster"
import { ImprovedCritical } from "../feats/shared/ImprovedCritical"
import { WeaponMasteries } from "../feats/shared/WeaponMasteries"
import { Character } from "../sim/Character"
import { ClassLevel } from "../sim/coreFeats/ClassLevel"
import { AttackRollEvent } from "../sim/events/AttackRollEvent"
import { BeginTurnEvent } from "../sim/events/BeginTurnEvent"
import { Feat } from "../sim/Feat"
import { WeaponMastery } from "../sim/types"
import { Weapon } from "../sim/Weapon"
import { applyFeatSchedule, defaultMagicBonus } from "../util/helpers"
import { SavageAttacker } from "../feats/origin/SavageAttacker"
import { IrresistibleOffense } from "../feats/epic/IrresistibleOffense"
import { Greatsword } from "../weapons/index"
import { Maul } from "../weapons/index"
import { AttackResultEvent } from "../sim/events/AttackResultEvent"
import { Environment } from "../sim/Environment"
import { CustomTurn } from "../sim/actions/CustomTurn"
import {
    AttackActionOperation,
    NumAttacksAttribute,
    WeaponAttack,
} from "../sim/actions/AttackAction"
import { Resource } from "../sim/resources/Resource"
import { Operation, TurnStage } from "../sim/actions/Operation"
import { ActionOperation } from "../sim/actions/ActionOperation"

const ActionSurgeResource = "ActionSurge"

class ActionSurgeOperation implements Operation {
    repeatable: boolean = false

    eligible(environment: Environment): boolean {
        return environment.character.getResource(ActionSurgeResource)!.has()
    }

    do(environment: Environment): void {
        environment.character.getResource(ActionSurgeResource)!.use()
        environment.character.actions.add(1, true)
    }
}

class AddActionSurge extends Feat {
    apply(character: Character): void {
        if (!character.resources.has(ActionSurgeResource)) {
            character.resources.set(
                ActionSurgeResource,
                new Resource({
                    name: ActionSurgeResource,
                    character,
                    initialMax: 1,
                    resetOnShortRest: true,
                })
            )
        }
        character.getResource(ActionSurgeResource).addMax(1)
    }
}

class StudiedAttacks extends Feat {
    enabled: boolean = false

    apply(character: Character): void {
        character.events.on("attack_roll", (event) => this.attackRoll(event))
        character.events.on("attack_result", (event) =>
            this.attackResult(event)
        )
    }

    attackRoll(event: AttackRollEvent): void {
        if (this.enabled) {
            event.adv = true
            this.enabled = false
        }
    }

    attackResult(event: AttackResultEvent): void {
        if (!event.hit) {
            this.enabled = true
        }
    }
}

class HeroicAdvantage extends Feat {
    used: boolean = false

    apply(character: Character): void {
        character.events.on("begin_turn", (event) => this.beginTurn(event))
        character.events.on("attack_roll", (event) => this.attackRoll(event))
    }

    beginTurn(event: BeginTurnEvent): void {
        this.character!.heroicInspiration.add(1)
    }

    attackRoll(event: AttackRollEvent): void {
        if (event.adv) {
            return
        }
        if (this.character!.heroicInspiration.has()) {
            const roll = event.roll1
            if (roll < 8) {
                this.character!.heroicInspiration.use()
                event.adv = true
            }
        }
    }
}

class PrecisionAttack extends Feat {
    low: number

    constructor(low: number) {
        super()
        this.low = low
    }

    apply(character: Character): void {
        character.events.on("attack_roll", (event) => this.attackRoll(event))
    }

    attackRoll(event: AttackRollEvent): void {
        if (
            event.attack.hasTag("used_maneuver") ||
            !this.character!.combatSuperiority.has() ||
            event.hits()
        ) {
            return
        }
        if (event.roll() >= this.low) {
            const roll = this.character!.combatSuperiority.roll()
            event.situationalBonus += roll
            event.attack.addTag("used_maneuver")
        }
    }
}

class TrippingAttack extends Feat {
    apply(character: Character): void {
        character.events.on("attack_result", (event) =>
            this.attackResult(event)
        )
    }

    attackResult(event: AttackResultEvent): void {
        if (
            !event.hit ||
            event.attack.target.prone ||
            event.attack.hasTag("used_maneuver") ||
            !this.character!.combatSuperiority.has()
        ) {
            return
        }
        const roll = this.character!.combatSuperiority.roll()
        event.addDamage({ source: "TrippingAttack", dice: [roll] })
        if (!event.attack.target.save(this.character!.dc("str"))) {
            event.attack.target.knockProne()
        }
    }
}

class CombatSuperiority extends Feat {
    level: number

    constructor(level: number) {
        super()
        this.level = level
    }

    apply(character: Character): void {
        let maxDice = 4
        if (this.level >= 15) {
            maxDice = 6
        } else if (this.level >= 7) {
            maxDice = 5
        }
        let die = 8
        if (this.level >= 18) {
            die = 12
        } else if (this.level >= 10) {
            die = 10
        }
        for (let i = 0; i < maxDice; i++) {
            character.combatSuperiority.addDie(die)
        }
    }
}

class Relentless extends Feat {
    apply(character: Character): void {
        character.combatSuperiority.enableRelentless()
    }
}

class ToppleIfNecessaryAttackAction extends ActionOperation {
    repeatable: boolean = true
    weapon: Weapon
    toppleWeapon: Weapon

    constructor(weapon: Weapon, toppleWeapon: Weapon) {
        super()
        this.weapon = weapon
        this.toppleWeapon = toppleWeapon
    }

    action(environment: Environment): void {
        const target = environment.target
        const character = environment.character
        const numAttacks = character.getAttribute(NumAttacksAttribute)
        for (let i = 0; i < numAttacks; i++) {
            let weapon = this.weapon
            if (!target.prone && i < numAttacks - 1) {
                weapon = this.toppleWeapon
            }
            character.weaponAttack({
                target,
                weapon,
                tags: ["main_action"],
            })
        }
    }
}

export class Fighter {
    static baseFeats(args: {
        level: number
        asis: Array<Feat>
        masteries: WeaponMastery[]
        fightingStyle: Feat
    }): Feat[] {
        const { level, asis, masteries, fightingStyle } = args
        const feats: Feat[] = []
        if (level >= 1) {
            feats.push(new ClassLevel("Fighter", level))
            feats.push(new WeaponMasteries(masteries))
            feats.push(fightingStyle)
        }
        if (level >= 2) {
            feats.push(new AddActionSurge())
        }
        if (level >= 5) {
            feats.push(new ExtraAttack(2))
        }
        if (level >= 11) {
            feats.push(new ExtraAttack(3))
        }
        if (level >= 13) {
            feats.push(new StudiedAttacks())
        }
        if (level >= 17) {
            feats.push(new AddActionSurge())
        }
        if (level >= 20) {
            feats.push(new ExtraAttack(4))
        }
        applyFeatSchedule({
            feats,
            newFeats: asis,
            schedule: [4, 6, 8, 12, 14, 16, 19],
            level,
        })
        return feats
    }

    static championFeats(level: number): Feat[] {
        const feats: Feat[] = []
        if (level >= 3) {
            feats.push(new ImprovedCritical(19))
        }
        if (level >= 10) {
            feats.push(new HeroicAdvantage())
        }
        if (level >= 15) {
            feats.push(new ImprovedCritical(18))
        }
        return feats
    }

    static battlemasterFeats(level: number): Feat[] {
        const feats: Feat[] = []
        if (level >= 3) {
            feats.push(new CombatSuperiority(level))
        }
        if (level >= 15) {
            feats.push(new Relentless())
        }
        return feats
    }

    static createBattlemasterFighter(level: number): Character {
        const character = new Character({
            stats: { str: 17, dex: 10, con: 10, int: 10, wis: 10, cha: 10 },
        })
        const weapon = new Greatsword({ magicBonus: defaultMagicBonus(level) })
        const toppleWeapon = new Maul({ magicBonus: defaultMagicBonus(level) })
        const feats = []
        feats.push(new SavageAttacker())
        feats.push(
            ...Fighter.baseFeats({
                level,
                asis: [
                    new GreatWeaponMaster(weapon),
                    new AbilityScoreImprovement("str"),
                    new AbilityScoreImprovement("con"),
                    new AbilityScoreImprovement("con"),
                    new AbilityScoreImprovement("con"),
                    new AbilityScoreImprovement("con"),
                    new IrresistibleOffense("str"),
                ],
                masteries: ["Topple", "Graze"],
                fightingStyle: new GreatWeaponFighting(),
            })
        )
        feats.push(...Fighter.battlemasterFeats(level))
        feats.push(new PrecisionAttack(8))
        character.customTurn.addOperation(
            "before_action",
            new ActionSurgeOperation()
        )
        character.customTurn.addOperation(
            "action",
            new ToppleIfNecessaryAttackAction(weapon, toppleWeapon)
        )
        feats.forEach((feat) => character.addFeat(feat))
        return character
    }

    static createChampionFighter(level: number): Character {
        const character = new Character({
            stats: { str: 17, dex: 10, con: 10, int: 10, wis: 10, cha: 10 },
        })
        const weapon = new Greatsword({ magicBonus: defaultMagicBonus(level) })
        const toppleWeapon = new Maul({ magicBonus: defaultMagicBonus(level) })
        const feats = []
        feats.push(new SavageAttacker())
        feats.push(
            ...Fighter.baseFeats({
                level,
                asis: [
                    new GreatWeaponMaster(weapon),
                    new AbilityScoreImprovement("str"),
                    new AbilityScoreImprovement("con"),
                    new AbilityScoreImprovement("con"),
                    new AbilityScoreImprovement("con"),
                    new AbilityScoreImprovement("con"),
                    new IrresistibleOffense("str"),
                ],
                masteries: ["Graze", "Topple"],
                fightingStyle: new GreatWeaponFighting(),
            })
        )
        feats.push(...Fighter.championFeats(level))
        character.customTurn.addOperation(
            "before_action",
            new ActionSurgeOperation()
        )
        character.customTurn.addOperation(
            "action",
            new ToppleIfNecessaryAttackAction(weapon, toppleWeapon)
        )
        feats.forEach((feat) => character.addFeat(feat))
        return character
    }
}
